### ISSUE TITLE
Middleware called once, twice or not at all depending on usage

### ISSUE THREAD

**Issue by dagnelies** [honojs/hono#687]
Hi,

in my app, I used the following pattern, which caused the middleware to be called **twice** on each request.

```ts
app.use('/myroute', myMiddleware)
app.route('/myroute', myGroupedHono)
```

By using the following pattern, the middleware is **not** called at all, which is surprising considering the previous case.

```ts
app.use('/myroute', myMiddleware)
app.get('/myroute/whatever', myHandler)
```

By using the following pattern, the middleware is called **once**, as expected:

```ts
app.use('/myroute/*', myMiddleware)
app.route('/myroute', myGroupedHono)
```

I think it would be good to make this routing/middleware behavior more consistent, or at least document the curlpits. 😉

**Comment by yusukebe** [comment]
Hi @dagnelies !

You are right! This is a confusing behavior.
I think it should be changed. I'll consider it later once I've sorted it out. Thanks!

**Comment by dagnelies** [comment]
Well, I don't know what's going on ...I wonder why but my middleware is not called at all now ...I could have sworn it was not like this previously.

```js
app.use('/me', requireAuth)
app.route('/me', me)
```

=> Calling `/me/profile` **fails**


```js
app.use('/me/*', requireAuth)
app.route('/me', me)
```

=> Calling `/me/profile` **fails**


```js
app.use('/*', requireAuth)
app.route('/me', me)
```

=> Calling `/me/profile` **works** ...and called once

It's very weird.

**Comment by dagnelies** [comment]
I still don't know what's going on, but something is definitely broken ...like really really broken ...depending on the order in which the use/route statements are listed, the behavior seem to change

**Comment by dagnelies** [comment]
I still don't know what's the problem, but intermixing use/routes like

```js
app.use('/a/*', middlewareA)
app.route('/a', a)
app.use('/b/*', middlewareB)
app.route('/b', b)
app.use('/c/*', middlewareC)
app.route('/c', c)
```
does not work properly. If you switch to everything to `use` first before adding any routes, it works as expected. Avoid like even a simple routing beforehand.

**Comment by yusukebe** [comment]
What is shown if you write `app.showRoutes()` at the bottom?

**Comment by yusukebe** [comment]
Hmm, the short snippets I wrote works as expected. Is this not good?

```ts
const app = new Hono()

const middlewareA: Handler = async (c, next) => {
  console.log('A called')
  await next()
}
const middlewareB: Handler = async (c, next) => {
  console.log('B called')
  await next()
}

const a = new Hono()
a.get('/foo', (c) => c.text('foo'))

const b = new Hono()
b.get('/bar', (c) => c.text('bar'))

app.use('/a/*', middlewareA)
app.route('/a', a)
app.use('/b/*', middlewareB)
app.route('/b', b)

app.showRoutes()

export default app
```

```
GET /a/foo ---> 'foo' + A called // expected
GET /b/bar ---> 'bar' + B called // expected
```

**Comment by dagnelies** [comment]
I do not know the exact combination which caused the problem. I just know that reordering the routes made the issue of "ignored middleware" reappear/disappear. I'll still try to invest a little bit more time to isolate an example, but it already took me lots of time to figure out the problem so far.

**Comment by dagnelies** [comment]
Working reproduction, just by adding the logger for example.

```
import { Handler, Hono } from "hono"
import { logger } from "hono/logger"

const app = new Hono()

app.use('*', logger())

const middlewareA: Handler = async (c, next) => {
  console.log('A called')
  await next()
}
const middlewareB: Handler = async (c, next) => {
  console.log('B called')
  await next()
}

const a = new Hono()
a.get('/foo', (c) => c.text('foo'))

const b = new Hono()
b.get('/bar', (c) => c.text('bar'))

app.use('/a/*', middlewareA)
app.route('/a', a)
app.use('/b/*', middlewareB)
app.route('/b', b)

app.showRoutes()

export default app
```

Output: `A called` present but `B called` missing
```
  <-- GET /a/foo
A called
  --> GET /a/foo 200 0ms
GET /a/foo 200 OK (2.38ms)
  <-- GET /b/bar
  --> GET /b/bar 200 0ms
GET /b/bar 200 OK (2.03ms)
```

**Comment by yusukebe** [comment]
@dagnelies 

This is really a bug in the router. Sorry I didn't notice that and thank you for doing the detailed investigation!
I've made issue #699 for it.

**Comment by yusukebe** [comment]
Hi @dagnelies 

Sorry for being late. The bug has been fixed ( thanks @usualoma ! ), and we've released the new version `v2.5.10`. I think it behaves as expected. Thank you!
